File: 백준/Gold/2285. 우체국/우체국.cc
#include <iostream>
#include <climits>
#include <algorithm>
using namespace std;

int n;
pair<long long, long long> arr[100001];
long long sum;

int main() {

	cin >> n;

	for (int i = 1; i <= n; i++) {
		cin >> arr[i].first >> arr[i].second;
		sum += arr[i].second;
	}

	sort(arr + 1, arr + 1 + n);

	long long ans = 0, minAns = LLONG_MAX, idx = 0;

	for (int i = 1; i <= n; i++) {
		long long left = ans;
		long long right = sum - ans - arr[i].second;
		ans += arr[i].second;
		if (abs(left - right) < minAns) {
			minAns = abs(left - right);
			idx = arr[i].first;
		}
	}

	cout << idx;
}

File: 백준/Gold/2285. 우체국/우체국.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static List<int[]> arr = new ArrayList<>();
	static int n;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt(br.readLine());

		long total = 0;
		for (int i = 0; i < n; i++) {
			int a, b;
			var st = new StringTokenizer(br.readLine());
			a = Integer.parseInt(st.nextToken());
			b = Integer.parseInt(st.nextToken());
			arr.add(new int[] {a, b});
			total += b;
		}

		arr.sort(Comparator.comparing(a -> a[0]));

		long left = 0, right = total;
		long ans = Long.MAX_VALUE;
		int idx = 0;
		for (int i = 0; i < n; i++) {
			right -= arr.get(i)[1];
			if (Math.abs(left - right) < ans) {
				ans = Math.abs(left - right);
				idx = arr.get(i)[0];
			}
			left += arr.get(i)[1];
		}

		System.out.println(idx);
	}
}